File: UnitTest/Zeze/Trans/TestTransactionLevelSerialiable.cs

using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Threading.Tasks;
using Zeze.Transaction;

namespace UnitTest.Zeze.Trans
{
    [TestClass]
    public class TestTransactionLevelSerialiable
    {
        [TestInitialize]
        public void TestInit()
        {
            demo.App.Instance.Start();
        }

        [TestCleanup]
        public void TestCleanup()
        {
            demo.App.Instance.Stop();
        }

        private volatile bool InTest = true;

        [TestMethod]
        public void Test2()
        {
            demo.App.Instance.Zz.NewProcedure(Init, "test_init").CallSynchronously();
            Task.Run(Verify_task);
            try
            {
                Task[] tasks = new Task[200000];
                for (int i = 0; i < tasks.Length; ++i)
                {
                    tasks[i] = demo.App.Instance.Zz.NewProcedure(Trade, "test_trade").CallAsync();
                }
                Task.WaitAll(tasks);
            }
            finally
            {
                InTest = false;
            }
        }

        private void Verify_task()
        {
            while (InTest)
            {
                demo.App.Instance.Zz.NewProcedure(Verify, "test_verify").CallSynchronously();
            }
        }

        private async Task<long> Verify()
        {
            var v1 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1L);
            var v2 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(2L);
            var total = v1.Int1 + v2.Int1;
            // 必须在事务成功时verify，执行过程中是可能失败的。
            Transaction.Current.RunWhileCommit(() => { Assert.AreEqual(100_000, total); });
            return 0;
        }

        private async Task<long> Init()
        {
            var v1 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1L);
            var v2 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(2L);
            v1.Int1 = 100_000;
            v2.Int1 = 0;
            return 0;
        }

        private async Task<long> Trade()
        {
            var v1 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1L);
            var v2 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(2L);
            var money = global::Zeze.Util.Random.Instance.Next(1000);
            if (global::Zeze.Util.Random.Instance.Next(100) < 50)
            {
                // random swap
                (v2, v1) = (v1, v2);
            }
            v1.Int1 -= money;
            v2.Int1 += money;
            return 0;
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Trans/TestTransactionLevelSerialiable.java
package UnitTest.Zeze.Trans;

import java.util.concurrent.Future;
import Zeze.Transaction.DispatchMode;
import Zeze.Transaction.Transaction;
import demo.App;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class TestTransactionLevelSerialiable {
	@Before
	public final void testInit() throws Exception {
		demo.App.getInstance().Start();
	}

	@After
	public final void testCleanup() throws Exception {
		//demo.App.getInstance().Stop();
	}

	private volatile boolean InTest = true;

	@Test
	public final void Test2() throws Exception {
		App.Instance.Zeze.newProcedure(TestTransactionLevelSerialiable::init, "test_init").call();
		Zeze.Util.Task.run(this::verify_task, "verify_task", DispatchMode.Normal);
		try {
			Future<?>[] tasks = new Future[20000];
			for (int i = 0; i < tasks.length; ++i) {
				tasks[i] = Zeze.Util.Task.runUnsafe(
						App.Instance.Zeze.newProcedure(TestTransactionLevelSerialiable::trade, "test_trade"),
						DispatchMode.Normal);
			}
			Zeze.Util.Task.waitAll(tasks);
		} finally {
			InTest = false;
		}
	}

	private void verify_task() throws Exception {
		while (InTest) {
			App.Instance.Zeze.newProcedure(TestTransactionLevelSerialiable::verify, "test_verify").call();
		}
	}

	private static long verify() {
		var v1 = App.Instance.demo_Module1.getTable1().getOrAdd(1L);
		var v2 = App.Instance.demo_Module1.getTable1().getOrAdd(2L);
		final var total = v1.getInt_1() + v2.getInt_1();
		// 必须在事务成功时verify，执行过程中是可能失败的。
		Transaction.whileCommit(() -> Assert.assertEquals(total, 100_000));
		return 0;
	}

	private static long init() {
		var v1 = App.Instance.demo_Module1.getTable1().getOrAdd(1L);
		var v2 = App.Instance.demo_Module1.getTable1().getOrAdd(2L);
		v1.setInt_1(100_000);
		v2.setInt_1(0);
		return 0;
	}

	private static long trade() {
		var v1 = App.Instance.demo_Module1.getTable1().getOrAdd(1L);
		var v2 = App.Instance.demo_Module1.getTable1().getOrAdd(2L);
		var money = Zeze.Util.Random.getInstance().nextInt(1000);
		if (Zeze.Util.Random.getInstance().nextBoolean()) {
			// random swap
			var tmp = v1;
			v1 = v2;
			v2 = tmp;
		}
		v1.setInt_1(v1.getInt_1() - money);
		v2.setInt_1(v2.getInt_1() + money);
		return 0;
	}
}
